File: lib/coin/weekly_mission_service.dart
// weekly_mission_service.dart — Nhiệm vụ tuần: chơi ít nhất 1 game/nhóm
//
// Reward: +500 xu khi hoàn thành cả 4 nhóm trong 1 tuần

import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../home/game_entry.dart';
import 'coin_service.dart';

// =============================================================================
// DATA MODEL
// =============================================================================

class WeeklyMissionData {
  final Set<GameCategory> categoriesDone;
  final bool categoryRewardClaimed;
  final int gamblingWinsThisWeek;
  final bool gamblingRewardClaimed;
  final String weekStartDate;

  const WeeklyMissionData({
    required this.categoriesDone,
    required this.categoryRewardClaimed,
    required this.gamblingWinsThisWeek,
    required this.gamblingRewardClaimed,
    required this.weekStartDate,
  });

  bool get allCategoriesDone => categoriesDone.length >= 4;
  int get categoriesCompleted => categoriesDone.length;
}

// =============================================================================
// WEEKLY MISSION SERVICE
// =============================================================================

class WeeklyMissionService {
  WeeklyMissionService._();
  static final WeeklyMissionService instance = WeeklyMissionService._();

  static const _kWeekStart        = 'weekly_week_start';
  static const _kCatsDone         = 'weekly_categories_done';
  static const _kCatRewardClaimed = 'weekly_cat_reward_claimed';
  static const _kGamblingWins     = 'weekly_gambling_wins';
  static const _kGamRewardClaimed = 'weekly_gam_reward_claimed';

  // ── In-memory ─────────────────────────────────────────────────────────────
  Set<GameCategory> _categoriesDone = {};
  bool _categoryRewardClaimed = false;
  int _gamblingWins = 0;
  bool _gamblingRewardClaimed = false;
  String _weekStart = '';

  final ValueNotifier<WeeklyMissionData> notifier = ValueNotifier(
    const WeeklyMissionData(
      categoriesDone: {},
      categoryRewardClaimed: false,
      gamblingWinsThisWeek: 0,
      gamblingRewardClaimed: false,
      weekStartDate: '',
    ),
  );

  // ── Helpers ───────────────────────────────────────────────────────────────

  /// Monday của tuần hiện tại (ISO 8601)
  String _currentWeekStart() {
    final now = DateTime.now().toLocal();
    final monday = now.subtract(Duration(days: now.weekday - 1));
    return '${monday.year}-${monday.month.toString().padLeft(2, '0')}-${monday.day.toString().padLeft(2, '0')}';
  }

  WeeklyMissionData _buildData() => WeeklyMissionData(
        categoriesDone: Set.unmodifiable(_categoriesDone),
        categoryRewardClaimed: _categoryRewardClaimed,
        gamblingWinsThisWeek: _gamblingWins,
        gamblingRewardClaimed: _gamblingRewardClaimed,
        weekStartDate: _weekStart,
      );

  void _notify() => notifier.value = _buildData();

  // ── Init ──────────────────────────────────────────────────────────────────

  Future<void> init() async {
    final prefs = await SharedPreferences.getInstance();
    final currentWeek = _currentWeekStart();
    final savedWeek = prefs.getString(_kWeekStart) ?? '';

    if (savedWeek != currentWeek) {
      // Tuần mới → reset
      _categoriesDone = {};
      _categoryRewardClaimed = false;
      _gamblingWins = 0;
      _gamblingRewardClaimed = false;
      _weekStart = currentWeek;

      await prefs.setString(_kWeekStart, currentWeek);
      await prefs.remove(_kCatsDone);
      await prefs.setBool(_kCatRewardClaimed, false);
      await prefs.setInt(_kGamblingWins, 0);
      await prefs.setBool(_kGamRewardClaimed, false);
    } else {
      _weekStart = savedWeek;
      _categoryRewardClaimed = prefs.getBool(_kCatRewardClaimed) ?? false;
      _gamblingRewardClaimed = prefs.getBool(_kGamRewardClaimed) ?? false;
      _gamblingWins = prefs.getInt(_kGamblingWins) ?? 0;

      try {
        final json = prefs.getString(_kCatsDone) ?? '[]';
        final list = (jsonDecode(json) as List).cast<String>();
        _categoriesDone = list
            .map((s) => GameCategory.values.firstWhere(
                  (c) => c.name == s,
                  orElse: () => GameCategory.puzzle,
                ))
            .toSet();
      } catch (_) {
        _categoriesDone = {};
      }
    }

    _notify();
  }

  // ── Record ────────────────────────────────────────────────────────────────

  /// Gọi sau mỗi lần user quay về home từ 1 game
  Future<void> recordGamePlayed(GameCategory category) async {
    if (_categoriesDone.contains(category)) return;

    final prefs = await SharedPreferences.getInstance();
    _categoriesDone.add(category);
    await prefs.setString(
        _kCatsDone, jsonEncode(_categoriesDone.map((c) => c.name).toList()));

    // Check category reward
    if (_categoriesDone.length >= 4 && !_categoryRewardClaimed) {
      _categoryRewardClaimed = true;
      await prefs.setBool(_kCatRewardClaimed, true);
      await CoinService.instance.earnCoins(500);
    }

    _notify();
  }

  /// Gọi khi thắng gambling game (Bầu Cua, Đỏ Đen, Xì Jack)
  Future<void> recordGamblingWin() async {
    final prefs = await SharedPreferences.getInstance();
    _gamblingWins++;
    await prefs.setInt(_kGamblingWins, _gamblingWins);

    // 3 lần thắng gambling trong tuần → +300 xu
    if (_gamblingWins >= 3 && !_gamblingRewardClaimed) {
      _gamblingRewardClaimed = true;
      await prefs.setBool(_kGamRewardClaimed, true);
      await CoinService.instance.earnCoins(300);
    }

    _notify();
  }
}

File: super_gate_web/src/services/weeklyMissionService.ts
import { CoinService } from './coinService';

export type GameCategory = 'lucky' | 'action' | 'puzzle' | 'strategy';

export interface WeeklyMissionData {
  categoriesDone: GameCategory[];
  categoryRewardClaimed: boolean;
  gamblingWinsThisWeek: number;
  gamblingRewardClaimed: boolean;
  weekStartDate: string;
  allCategoriesDone: boolean;
  categoriesCompleted: number;
}

export class WeeklyMissionService {
  private static readonly KEYS = {
    weekStart: 'weekly_week_start',
    catsDone: 'weekly_categories_done',
    catRewardClaimed: 'weekly_cat_reward_claimed',
    gamblingWins: 'weekly_gambling_wins',
    gamRewardClaimed: 'weekly_gam_reward_claimed',
  };

  private static categoriesDone: Set<GameCategory> = new Set();
  private static categoryRewardClaimed = false;
  private static gamblingWins = 0;
  private static gamblingRewardClaimed = false;
  private static weekStart = '';

  private static listeners: ((data: WeeklyMissionData) => void)[] = [];

  static subscribe(listener: (data: WeeklyMissionData) => void) {
    this.listeners.push(listener);
    listener(this.getData());
    return () => {
      this.listeners = this.listeners.filter((l) => l !== listener);
    };
  }

  private static notify() {
    const data = this.getData();
    this.listeners.forEach((listener) => listener(data));
  }

  private static getMondayStr(): string {
    const now = new Date();
    const day = now.getDay();
    const diff = now.getDate() - day + (day === 0 ? -6 : 1); // Adjust when Sunday
    const monday = new Date(now.setDate(diff));
    
    const year = monday.getFullYear();
    const month = String(monday.getMonth() + 1).padStart(2, '0');
    const dStr = String(monday.getDate()).padStart(2, '0');
    return `${year}-${month}-${dStr}`;
  }

  static init() {
    const currentWeek = this.getMondayStr();
    const savedWeek = localStorage.getItem(this.KEYS.weekStart) || '';

    if (savedWeek !== currentWeek) {
      // Tuần mới -> Reset toàn bộ
      this.categoriesDone = new Set();
      this.categoryRewardClaimed = false;
      this.gamblingWins = 0;
      this.gamblingRewardClaimed = false;
      this.weekStart = currentWeek;

      localStorage.setItem(this.KEYS.weekStart, currentWeek);
      localStorage.setItem(this.KEYS.catsDone, JSON.stringify([]));
      localStorage.setItem(this.KEYS.catRewardClaimed, 'false');
      localStorage.setItem(this.KEYS.gamblingWins, '0');
      localStorage.setItem(this.KEYS.gamRewardClaimed, 'false');
    } else {
      this.weekStart = savedWeek;
      this.categoryRewardClaimed = localStorage.getItem(this.KEYS.catRewardClaimed) === 'true';
      this.gamblingRewardClaimed = localStorage.getItem(this.KEYS.gamRewardClaimed) === 'true';
      this.gamblingWins = parseInt(localStorage.getItem(this.KEYS.gamblingWins) || '0');

      try {
        const arr = JSON.parse(localStorage.getItem(this.KEYS.catsDone) || '[]');
        this.categoriesDone = new Set(arr);
      } catch (_) {
        this.categoriesDone = new Set();
      }
    }
    this.notify();
  }

  static getData(): WeeklyMissionData {
    const categoriesDone = Array.from(this.categoriesDone);
    return {
      categoriesDone,
      categoryRewardClaimed: this.categoryRewardClaimed,
      gamblingWinsThisWeek: this.gamblingWins,
      gamblingRewardClaimed: this.gamblingRewardClaimed,
      weekStartDate: this.weekStart,
      allCategoriesDone: this.categoriesDone.size >= 4,
      categoriesCompleted: this.categoriesDone.size,
    };
  }

  static async recordGamePlayed(category: GameCategory): Promise<void> {
    this.init(); // Đảm bảo đúng tuần trước
    if (this.categoriesDone.has(category)) return;

    this.categoriesDone.add(category);
    localStorage.setItem(this.KEYS.catsDone, JSON.stringify(Array.from(this.categoriesDone)));

    // Chơi đủ 4 loại game -> thưởng 500 xu
    if (this.categoriesDone.size >= 4 && !this.categoryRewardClaimed) {
      this.categoryRewardClaimed = true;
      localStorage.setItem(this.KEYS.catRewardClaimed, 'true');
      await CoinService.earnCoins(500);
    }

    this.notify();
  }

  static async recordGamblingWin(): Promise<void> {
    this.init(); // Đảm bảo đúng tuần trước
    this.gamblingWins++;
    localStorage.setItem(this.KEYS.gamblingWins, String(this.gamblingWins));

    // Thắng 3 ván cược -> thưởng 300 xu
    if (this.gamblingWins >= 3 && !this.gamblingRewardClaimed) {
      this.gamblingRewardClaimed = true;
      localStorage.setItem(this.KEYS.gamRewardClaimed, 'true');
      await CoinService.earnCoins(300);
    }

    this.notify();
  }
}

// Khởi chạy ngay khi load module
WeeklyMissionService.init();
